### ISSUE TITLE
Please implement "show property on panel icon" feature.

### ISSUE THREAD

**Issue by ikunya** [fcitx/fcitx#245]
This is important for Japanese.

IBus is already done it.

https://github.com/ibus/ibus/commit/23c45b970b195008a54884a1a9d810e7f8b22c5c

https://desktopi18n.wordpress.com/2015/07/21/ibus-1-5-11-is-released/

**Comment by wengxt** [comment]
what a ugly hack!...

I have some other idea to present it in a different way.

Have you notice that fcitx can show current im name when you switch im? I think we may add some additional im specific mode information there.

**Comment by ikunya** [comment]
Fcitx shows hint when user switches im. I think Fcitx needs not to show current im on panel icon.
I don't know this is the answer what you want.

**Comment by wengxt** [comment]
well, if I understand it correctly, the purpose of this functionality is to let user know what's the sub mode of current IM. Correct?

It doesn't really need to be in the tray icon, or in the same icon.

As for the ugly hack I mean that animation implemented in ibus. I don't think it is the best way to implement such feature.

so far, the different ways to present such information that I can think of are:
1. use system notification
2. use a extra tray icon
3. show some text in the input panel (where the candidates shows), eg, currently we show "Anthy" when switch im, we may change it to show "Anthy - Hiragana".
4. show in the same tray icon, but with better icon design.

fcitx have more than one ui implementation, there are certain feature could not be implemented in all ui. Currently, the panel icon for anthy's mode is drawn with text. SNI based tray icon cannot show arbitrary pixmap. So if we want icon we need to draw more.

And kimpanel is a different story, kde shows all icon, while the gnome one shows only one. If we need extra icon we probably want some hint information. Also, more than one xembed icon per app may result in bad icon order in tray.

That's why I perfer method 3 because it's generally work for all ui.

what do you think? regardless of implementation details, can you tell me what do you think about those 4 methods I propose?

**Comment by ikunya** [comment]
> well, if I understand it correctly, the purpose of this functionality is to let user know what's the sub mode of current IM. Correct?

Yes.

I would like to say that I want to know the current status of im without nothing to do if it can do.

My currently setting is to show the old-fashioned status panel. This is always shown the current im status whenever I want to know. But this is deprecated. So sad.

Method 3 looks like IBus's property panel.

![virtualbox_kubuntuwily2_21_02_2016_13_11_42](https://cloud.githubusercontent.com/assets/11596277/13200846/da5d8f72-d89c-11e5-963e-2eda1eacadd5.png)

This is not bad idea. When it sets auto-hide mode (likes currently im hint), no way to know the current im mode. It is annoying. I think it needs one more to show the current im status. eg. put Ctrl+Alt+M and shows the current status. May be it is method 1.

So my opinion is method 3 and 1.

**Comment by wengxt** [comment]
There is a certain reason that 3 is better than other three ways. (not to mention that we already have such a mechanism and it would be easy to implement and it will be available for all fcitx's ui).

It shows at current cursor position so you don't need to look around. Method 1 may be too annoying to users, currently I only use it when hotkey is used to trigger a infrequently used feature.

I could certainly move the fcitx's panel together with the candidate window, but the positioning and when to hide the window could also be a problem. I don't see any relate code to deal with overlapping of candidate panel and property panel in ibus, probably they just ignore such a thing. And also current two of kimpanel implementations doesn't allow you to have a floating panel for property.

BTW, I don't event think the ibus ui under gnome-shell support that.

**Comment by ikunya** [comment]
I am looking forward to the implementation!

 > BTW, I don't event think the ibus ui under gnome-shell support that.

GNOME Shell intends to show the property on panel icon.

**Comment by wengxt** [comment]
I mean the same feature you request in this issue is not supported by gnome-shell's ibus ui (which is in gnome-shell's repo). icon_prop_key is currently not used in gnome-shell's ibus ui implementation.

Anyway that's off topic.

**Comment by wengxt** [comment]
Please check the video, does it look ok for you?

[![Video](http://img.youtube.com/vi/w3JtkhCbUBs/0.jpg)](https://www.youtube.com/watch?v=w3JtkhCbUBs)

**Comment by ikunya** [comment]
GNOME Shell itself has such an icon instead of IBus,

**Comment by ikunya** [comment]
Thank you for your work. It looks good to me.

And how do I notice the current mode? Changes the focus or im modes?

**Comment by wengxt** [comment]
That information is shown when mode is changed, or you change your focus to another application , or you change your im to anthy from something else. 

I could also make im has a alternative icon to show, but really we would need some good icon before put it into real usage (some icon that make people know it's anthy in another mode).

**Comment by ikunya** [comment]
With my understanding, we need 5 (_A/あ/ア/ｱ/Ａ) icons for Anthy. Is it correct?

**Comment by wengxt** [comment]
Yeah, the real problem is the icon design. For example, we could just use plain あ as icon, but what about mozc or kkc? (IMHO mozc's icon is better for this case, because it has a orange-ish background). Also current version of anthy's icon also looks not modern enough.

A good example in pinyin world is https://github.com/fcitx/fcitx-sunpinyin/blob/master/data/sunpinyin-logo-128.png. As you can see, the sunflower lets you know it is "sunpinyin" (the origin of sun is Sun Inc and probably also its authers first name), also it has 拼 in it.

I'm thinking of create a new icon set by combining the old crown icon with symbol of input mode in the same way. The name of anthy originate from a character with crown anyway.

**Comment by ikunya** [comment]
Any progress?

And I want to use current show input mode from another input methods especially for fcitx-skk.

I don't know much about skk, but some experienced skk users say the same thing.

Would you please implement it?
